feature 533 - Add deg2pix

OmerShubi opened the pull request:
## Description

Adds `deg2pix` method. 
Fixes issue #533.

## Implemented changes

Added a deg2pix method.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change is or requires a documentation update

## How Has This Been Tested?

Added the same tests as in `tests/unit/gaze/transforms/pix2deg_test.py` with the following changes:
- Replaced pixel<-->position in output<-->input. 
- Added atol=1-e4 in equality assertions. @dkrako Is this okay? Alternative is to refine the input values.
- Added two digits after zero for input position value in isosceles_triangle_origin_lowerleft_returns_45 and isosceles_triangle_left_origin_lowerleft_returns_neg45.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

OmerShubi commented:
I didn't want to make too many changes but I used the distance conversion from cm to px as is. 
Can consider extracting a function like:

```python
def distance_cm_to_px(
        distance: float | str,
        screen_resolution: tuple[int, int],
        screen_size: tuple[float, float],
        n_components: int,) -> pl.Expr:
    
    
    if isinstance(distance, (float, int)):
        _check_distance(distance)
        distance_series = pl.lit(distance)
    elif isinstance(distance, str):
        # True division by 10 is needed to convert distance from mm to cm
        distance_series = pl.col(distance).truediv(10)
    else:
        raise TypeError(
            f'`distance` must be of type `float`, `int` or `str`, but is of type'
            f'`{type(distance).__name__}`',
        )
    
    distance_pixels = pl.concat_list([
        distance_series.mul(screen_resolution[component % 2] / screen_size[component % 2])
        for component in range(n_components)
    ])
    
    return distance_pixels
```

OmerShubi commented:
I tried adding a test for deg2pix in GazeDataFrame but am getting 
`FAILED tests/unit/gaze/apply_test.py::test_gaze_apply[deg2pix_origin_center] - AssertionError: DataFrames are different (columns are not in the same order)`. 
I think that the column order for the expected df is defined here
https://github.com/aeye-lab/pymovements/blob/b097c93e966e8f0dbdc460708a77b692f3da3149/src/pymovements/gaze/gaze_dataframe.py#L267-L275
and in `deg2pix` it appends the pixel column after position. 
I'm not sure how to go about this. @dkrako any suggestions?

dkrako commented:
> I tried adding a test for deg2pix in GazeDataFrame but am getting `FAILED tests/unit/gaze/apply_test.py::test_gaze_apply[deg2pix_origin_center] - AssertionError: DataFrames are different (columns are not in the same order)`. I think that the column order for the expected df is hardcoded in the creation of the GazeDataFrame and in `deg2pix` it appends the pixel column after position.
> 
> I'm not sure how to go about this. @dkrako any suggestions?

Thank you for checking in before changing the existing testing.
I thought about it a bit and I think it's safe to add `check_column_order=True` when calling `assert_frame_equal()`.

I don't think that it's necessary to enforce a column ordere here and the different column ordering makes sense actually:

1. when we have only a `position` column and add a `pixel` column via `deg2pix()` I would expect a FIFO column order.
2. when we create a new `GazeDataFrame` the columns are in a default initialization order.

OmerShubi commented:
> Thank you for checking in before changing the existing testing. I thought about it a bit and I think it's safe to add `check_column_order=True` when calling `assert_frame_equal()`.

You mean `check_column_order=False` right?

Anyways, added a suggested fix.

dkrako reviewed (changes requested):
This is excellent! Thank you so much for the implementation.

There are some very minor changes I would like to request which shouldn't take you long.

We'll discuss the naming of `origin` vs `pixel_origin` in `deg2pix()` on our friday meeting.

OmerShubi commented on src/pymovements/gaze/transforms.py:
```diff
@@ -276,15 +276,99 @@ def pix2deg(
+        for component in range(n_components)
+    ])
+
+    pixel_components = [
```
Done

OmerShubi commented on src/pymovements/gaze/transforms.py:
```diff
@@ -276,15 +276,99 @@ def pix2deg(
+            f' Valid values are: {supported_origins}',
+        )
+
+    centered_pixels = pl.concat_list(
```
Done.
Thanks for the comment, I agree the names were misleading.

OmerShubi commented:
@dkrako I added deg2pix tests in `deg2pix_test` and `apply_test` but I see also pix2deg tests in `dataset_processing_test.py`, `gaze_file_processing_test`, `public_dataset_processing_test`, 'dataset_test', `gaze_transform_test`, `transforms_library_test` and `traceplot_test` and `tsplot_test`

OmerShubi commented:
@dkrako I added deg2pix tests in `deg2pix_test` and `apply_test` but I see also pix2deg tests in 

- `dataset_processing_test.py`
- `gaze_file_processing_test`
- `public_dataset_processing_test`
- `dataset_test` 
- `gaze_transform_test`
- `transforms_library_test` 
- and `traceplot_test` and `tsplot_test`

Where would be relevant places to add deg2pix tests?

OmerShubi commented:
Also, renaming origin -> pixel_origin leads to failed test:
```        pytest.param(
            'deg2pix',
            {},
            pm.GazeDataFrame(
                data=pl.from_dict(
                    {
                        'time': [1000, 1000],
                        'x_dva': [26.335410003881346, 26.335410003881346],
                        'y_dva': [0.0, 0.0],
                    },
                ),
                experiment=pm.Experiment(100, 100, 100, 100, 100, 'center', 1000),
                position_columns=['x_dva', 'y_dva'],
            ),
```
Should I make changes in transform (see below)? or to the test? I am not sure how to handle this.
https://github.com/aeye-lab/pymovements/blob/53aa81926f91ded4990dc04906e0fd40e41cefd7/src/pymovements/gaze/gaze_dataframe.py#L351-L354

dkrako commented:
Regarding the missing tests, these three are missing:

- [ ] `test_gaze_transform_expected_frame()` in `gaze_transforms_test.py` misses a few cases: add deg2pix equivalents of the ids `pix2deg`, `pix2deg_origin_center`, `pix2deg_origin_lower_left` and `pix2deg_origin_center_distance_column`

- [ ] `test_gaze_dataframe_deg2pix_exceptions()` in `gaze_transfom_test.py`: add an deg2pix equivalent of this to check exceptions: https://github.com/aeye-lab/pymovements/blob/53aa81926f91ded4990dc04906e0fd40e41cefd7/tests/unit/gaze/gaze_transform_test.py#L778-L847

- [ ] `test_deg2pix()` in `dataset_test.py`, similar to this: https://github.com/aeye-lab/pymovements/blob/53aa81926f91ded4990dc04906e0fd40e41cefd7/tests/unit/dataset/dataset_test.py#L536-L548

As the `dataset_configuration` fixture already has a pixel column, let's adjust the new test to something like this:

```python
dataset.pix2deg()
dataset.deg2pix(output_column='new_pixel')

expected_schema = {**original_schema, 'new_pixel': pl.List(pl.Float64)} 
for result_gaze_df in dataset.gaze: 
    assert result_gaze_df.schema == expected_schema 
```

You don't need to check the values for correctness at the dataset level.

dkrako commented:
> Also, renaming origin -> pixel_origin leads to failed test:
> 
> ```
>             'deg2pix',
>             {},
>             pm.GazeDataFrame(
>                 data=pl.from_dict(
>                     {
>                         'time': [1000, 1000],
>                         'x_dva': [26.335410003881346, 26.335410003881346],
>                         'y_dva': [0.0, 0.0],
>                     },
>                 ),
>                 experiment=pm.Experiment(100, 100, 100, 100, 100, 'center', 1000),
>                 position_columns=['x_dva', 'y_dva'],
>             ),
> ```
> 
> Should I make changes in transform (see below)? or to the test? I am not sure how to handle this.
> 
> https://github.com/aeye-lab/pymovements/blob/53aa81926f91ded4990dc04906e0fd40e41cefd7/src/pymovements/gaze/gaze_dataframe.py#L351-L354

Would this be resolved if we set the default `pixel_origin: str = 'lower left'` in `transforms.py::deg2pix()`?

OmerShubi commented:
> * [x]  `test_gaze_transform_expected_frame()` in `gaze_transforms_test.py` misses a few cases: add deg2pix equivalents of the ids `pix2deg`, `pix2deg_origin_center`, `pix2deg_origin_lower_left` and `pix2deg_origin_center_distance_column`

Okay, added the missing tests, except an equivalent for `pix2deg` which I did not find. 




> Would this be resolved if we set the default `pixel_origin: str = 'lower left'` in `transforms.py::deg2pix()`?

Not really, but it seems reasonable to add it so I added it. The main issue, which I solved by passing the parameter directly, 
 was that the experiment doesn't have a `pixel_origin` parameter. Maybe this is okay.

OmerShubi commented on tests/unit/gaze/transforms/deg2pix_test.py:
```diff
@@ -0,0 +1,647 @@
+            },
+            pl.Series('position', [[0, 45]], pl.List(pl.Float64)),
+            pl.Series('pixel', [[0, 50]], pl.List(pl.Float64)),
+            id='isosceles_triangle_origin_center_returns_45',
```
Done

OmerShubi commented:
I think the latest commit covers all the comments you had @dkrako

dkrako reviewed (approved):
Great! Thank you so much for putting effort into the issue, clean code and quick development!
